plutus-core/testlib/PlutusCore/Test.hs

```diff
@@ -423,47 +425,145 @@ brokenRename ::
   m t
 brokenRename mark renM = through mark >=> runBrokenRenameT . renM
 
+{- Note [Scoping tests API]
+If you want to test how a certain pass handles scoping you should use either 'test_scopingGood' or
+'test_scopingBad' depending on whether the pass is expected to preserve global uniqueness and not
+change scoping of its argument. Regarding the last one: substitution, for example, may remove free
+variables and scoping tests, being initially developed only to test renamers, will fail upon seeing
+that a free variable was removed. So any scoping test failure needs to be carefully scrutinized
+before concluding that it reveals a bug.
+
+As it turns out, most AST transformations don't do anything that would cause the scoping tests to
+false-positively report a bug:
+
+- a lot of passes simply do not do anything with names apart from maybe moving them around
+- those that do change names may produce alpha-equivalent results and this is one thing that the
+  scoping machinery is designed to test
+- some passes such as inlining may duplicate binders, but that is also fine as long as the
+  duplicates are properly renamed, since the scoping machinery doesn't count binders or variable
+  usages, it only expects free names to stay free, bound names to change together with their binders
+  and global uniqueness to be preserved (see 'ScopeError' for the full list of possible errors)
+- some passes such as inlining may remove bindings and there's special support implemented for
+  handling this: when invoking either 'test_scopingGood' or 'test_scopingBad' you need to provide a
+  'BindingRemoval' argument specifying whether binding removal is expected for the pass. Conversily,
+  make it 'BindingRemovalOk' whenever you use 'test_scopingBad' to emphasize that even with binding
+  removal allowed tests still fail
+- some passes do not perform renaming of the input themselves, in that case you need to provide
+  'PrerenameYes' for the 'Prerename' argument that both the test runners expect. It doesn't matter
+  whether the pass relies on global uniqueness itself, because the scoping tests rely on it anyway.
+  If the pass renames its input, and only in this case, provide 'PrerenameNo' for the 'Prerename'
+  argument, this will allow the scoping tests to ensure that the pass does indeed rename its input
```

Comment:
Still think it would be easy enough to just test `rename . pass` in that case ...

Reply:
Only until you get a failure and need to understand what's going on. This whole thing is written in blood and tears.